File: Java/154.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public int findMin(int[] nums) {
        int start = 0, end = nums.length - 1;
        while (start < end) {
            int mid = (start + end) / 2;
            if (nums[mid] > nums[end]) {
                start = mid + 1;
            } else if (nums[mid] < nums[end]) {
                end = mid;
            } else {
                end--;
            }
        }
        return nums[start];
    }
}
__________________________________________________________________________________________________
sample 34640 kb submission
class Solution {
    public int findMin(int[] nums) {
        int l = 0, h = nums.length - 1;
        while(l < h){
            int mid = l + (h - l)/2;
            System.out.println(l+" "+mid+" "+h);
            if(nums[h] > nums[mid]){// rotation in low to mid
                h = mid;
            }
            else if(nums[mid] > nums[h]){
                l = mid+1;
            }else{
                h--;
            }
        }
        return Math.min(nums[l],nums[h]);
    }
}
__________________________________________________________________________________________________

File: Python3/154.py
__________________________________________________________________________________________________
sample 48 ms submission
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l, r = 0, len(nums)-1
        while l+1 < r:
            mid = (l+r)//2
            if nums[mid] == nums[l]:
                l += 1
            elif nums[mid] <= nums[r]:
                r = mid
            else:
                l = mid 
        if nums[l] < nums[r]:
            return nums[l] 
        else:
            return nums[r]        
__________________________________________________________________________________________________
sample 13216 kb submission
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if len(nums)==1:
            return nums[0]
        
        low, high = 0, len(nums)
        mid = (low + high) // 2
        if nums[low] == nums[high-1]:
            return min(self.findMin(nums[low:mid]),self.findMin(nums[mid:high]))
        
        if nums[low] < nums[high-1]:
            return nums[low]
        
        if nums[low] > nums[high-1]:
            if nums[mid]<nums[high-1]:
                return self.findMin(nums[low:mid+1])
            if nums[mid]==nums[high-1]:
                return min(self.findMin(nums[low:mid]),self.findMin(nums[mid:high]))            
            else:
                return self.findMin(nums[mid:high])
__________________________________________________________________________________________________
